include/seqan3/range/view/take_until.hpp

```diff
@@ -0,0 +1,449 @@
+     *                     seqan3::reference_t must model seqan3::common_reference_concept with `reference`.
+     * \returns This view converted to container_t.
+     */
+    template <container_concept container_t>
+    operator container_t()
+    //!\cond
+        requires common_reference_concept<reference_t<container_t>, reference>
+    //!\endcond
+    {
+        container_t ret;
+        ranges::copy(begin(), end(), ranges::back_inserter(ret));
+        return ret;
+    }
+};
+
+//!\brief Type deduction guide that strips references.
+//!\relates seqan3::detail::view_take_until
+template <typename urng_t, typename fun_t, bool or_throw = false>
+view_take_until(urng_t, fun_t) -> view_take_until<std::remove_reference_t<urng_t>, fun_t, or_throw>;
+
+// ============================================================================
+//  take_until_fn (adaptor definition)
+// ============================================================================
+
+/*!\brief View adaptor definition for view::take_until and view::take_until_or_throw.
+ * \tparam or_throw Whether to throw an exception when the input is exhausted before the end of line is reached.
+ */
+template <bool or_throw>
+class take_until_fn : public pipable_adaptor_base<take_until_fn<or_throw>>
+{
+private:
+    //!\brief Type of the CRTP-base.
+    using base_t = pipable_adaptor_base<take_until_fn<or_throw>>;
+
+public:
+    //!\brief Inherit the base class's Constructors.
+    using base_t::base_t;
+
+private:
+    //!\brief Befriend the base class so it can call impl().
+    friend base_t;
+
+    /*!\brief Call the view's constructor with the given parameters.
+     * \tparam    urng_t The underlying range type; must model seqan3::view_concept.
+     * \tparam    fun_t  The type of the callable; concept checks done in class.
+     * \param[in] urange The underlying range.
+     * \param[in] fun    The callable that will be evaluated on every element.
+     * \returns An instance of seqan3::detail::view_take_until.
+     */
+    template <view_concept urng_t, typename fun_t>
+    static auto impl(urng_t urange, fun_t && fun)
+    {
+        return view_take_until<urng_t, fun_t, or_throw>{std::move(urange), std::forward<fun_t>(fun)};
+    }
+
+    /*!\brief Wraps the range argument in seqan3::view::all and forwards to the other overload.
+     * \tparam    urng_t The underlying range type; must model seqan3::viewable_range_concept (be lvalue reference
+     *                   to non-view range).
+     * \tparam    fun_t  The type of the callable; concept checks done in class.
+     * \param[in] urange The underlying range.
+     * \param[in] fun    The callable that will be evaluated on every element.
+     * \returns An instance of seqan3::detail::view_take_until.
+     */
+    template <viewable_range_concept urng_t, typename fun_t>
+    static auto impl(urng_t && urange, fun_t && fun)
+    {
+        return impl(view::all(std::forward<urng_t>(urange)), std::forward<fun_t>(fun));
```

Comment:
Ok, that is an interesting turn: Now, what would be more efficient, adding another indirection layer using view::all or storing the urange with the ranges::sentinel_t<...> mf? I expect the sentinel thing needs to do some runtime checking to be used safely.  But do we know, whether the compiler noops the view::all access?